Error while connecting to openHAB REST API

VS code extension given an error  "Error while connecting to openHAB REST API". File editing works fine (via samba share).

Versions:
```
Visual Studio code: 1.18.1
openHAB VS Code Extension: 0.3.0
OS: macOS 10.13 High Sierra

openHAB: 2.2.0~20171018133740-1
OS: Linux version 4.4.0-96-generic
```

settings.json
```
{
    "files.associations": {
        "*.log": "properties"
    },
    "workbench.iconTheme": "openhab",
    "openhab.host": "192.168.10.7",
    "openhab.port": 8080
}
```

Wireshark:
192.168.10.22 = Visual Studio
192.168.10.7 = openHAB
![screen shot 2017-11-18 at 11 51 54](https://user-images.githubusercontent.com/5506533/32979065-ede2ab72-cc56-11e7-81b2-5345258747dd.png)

VS developer tools shows two errors

```
messageService.ts:126 Error while connecting to openHAB REST API. 
e.doShow @ messageService.ts:126
e.show @ messageService.ts:105
e._showMessage @ mainThreadMessageService.ts:72
n.Class.derive._oncancel @ winjs.base.raw.js:1641
e._showMessage @ mainThreadMessageService.ts:42
e.$showMessage @ mainThreadMessageService.ts:36
e.invoke @ abstractThreadService.ts:41
e._invokeHandler @ rpcProtocol.ts:109
e._receiveOneMessage @ rpcProtocol.ts:96
(anonymous) @ rpcProtocol.ts:32
(anonymous) @ rpcProtocol.ts:157
e.invoke @ callbackList.ts:36
e.fire @ event.ts:123
(anonymous) @ ipc.net.ts:82
emitOne @ events.js:96
emit @ events.js:191
readableAddChunk @ _stream_readable.js:178
Readable.push @ _stream_readable.js:136
onread @ net.js:560
```

```
shell.ts:459 Please return an array of children.: Error: Please return an array of children.
    at file:///Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:9:236755
    at Object.g [as _notify] (file:///Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:9:62701)
    at Object.enter (file:///Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:9:66048)
    at n.Class.derive._oncancel._run (file:///Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:9:67876)
    at n.Class.derive._oncancel._completed (file:///Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:9:67317)
    at e.resolveOk (file:///Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:9:2438440)
    at e._receiveOneMessage (file:///Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:9:2439971)
    at file:///Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:9:2439283
    at file:///Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:9:2441202
    at e.invoke (file:///Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:9:82095)
e.onUnexpectedError @ shell.ts:459
(anonymous) @ shell.ts:419
e.onUnexpectedError @ errors.ts:119
o @ errors.ts:138
v @ winjs.base.raw.js:1209
enter @ winjs.base.raw.js:901
_run @ winjs.base.raw.js:1068
_chainedError @ winjs.base.raw.js:1031
n @ winjs.base.raw.js:736
then @ winjs.base.raw.js:1436
enter @ winjs.base.raw.js:747
_run @ winjs.base.raw.js:1068
_setCompleteValue @ winjs.base.raw.js:1052
g @ winjs.base.raw.js:1170
enter @ winjs.base.raw.js:852
_run @ winjs.base.raw.js:1068
_completed @ winjs.base.raw.js:1036
e.resolveOk @ lazyPromise.ts:65
e._receiveOneMessage @ rpcProtocol.ts:72
(anonymous) @ rpcProtocol.ts:32
(anonymous) @ rpcProtocol.ts:157
e.invoke @ callbackList.ts:36
e.fire @ event.ts:123
(anonymous) @ ipc.net.ts:82
emitOne @ events.js:96
emit @ events.js:191
readableAddChunk @ _stream_readable.js:178
Readable.push @ _stream_readable.js:136
onread @ net.js:560

```

@kubawolanin, just tried the latest version (also reinstalled VSCode). Items works fine, but things panel stays empty.

Console:
```
messageService.ts:126 Error while connecting to openHAB REST API. 
e.doShow @ messageService.ts:126
e.show @ messageService.ts:105
e._showMessage @ mainThreadMessageService.ts:72
n.Class.derive._oncancel @ winjs.base.raw.js:1641
e._showMessage @ mainThreadMessageService.ts:42
e.$showMessage @ mainThreadMessageService.ts:36
e.invoke @ abstractThreadService.ts:41
e._invokeHandler @ rpcProtocol.ts:109
e._receiveOneMessage @ rpcProtocol.ts:96
(anonymous) @ rpcProtocol.ts:32
(anonymous) @ rpcProtocol.ts:157
e.invoke @ callbackList.ts:36
e.fire @ event.ts:123
(anonymous) @ ipc.net.ts:82
emitOne @ events.js:96
emit @ events.js:191
readableAddChunk @ _stream_readable.js:178
Readable.push @ _stream_readable.js:136
onread @ net.js:560

```

--- a/src/ItemsExplorer/ItemsExplorer.ts
+++ b/src/ItemsExplorer/ItemsExplorer.ts
@@ -3,8 +3,7 @@ import {
     EventEmitter,
     TreeDataProvider,
     TreeItem,
-    TreeItemCollapsibleState,
-    workspace
+    TreeItemCollapsibleState
 } from 'vscode'
 
 import { Item } from './Item'
@@ -73,8 +72,7 @@ export class ItemsExplorer implements TreeDataProvider<Item> {
     }
 
     public getChildren(item?: Item): Item[] | Thenable<Item[]> {
-        let config = workspace.getConfiguration('openhab')
-        if (!item && config.useRestApi) {
+        if (!item) {
             if (!this.model) {
                 this.model = new ItemsModel(this.openhabHost)
             }

--- a/src/ItemsExplorer/ItemsModel.ts
+++ b/src/ItemsExplorer/ItemsModel.ts
@@ -65,9 +65,7 @@ export class ItemsModel {
                     resolve(this.sort(transform(response)))
                 }.bind(this))
                 .catch(err => {
-                    if( err ) {
-                        reject()
-                    }
+                    handleRequestError(err).then(err => resolve([]))
                 })
         })
     }

--- a/src/ItemsExplorer/ItemsModel.ts
+++ b/src/ItemsExplorer/ItemsModel.ts
@@ -66,7 +66,6 @@ export class ItemsModel {
                 }.bind(this))
                 .catch(err => {
                     if( err ) {
-                        handleRequestError(err)
                         reject()
                     }
                 })

--- a/src/ItemsExplorer/ItemsModel.ts
+++ b/src/ItemsExplorer/ItemsModel.ts
@@ -65,8 +65,10 @@ export class ItemsModel {
                     resolve(this.sort(transform(response)))
                 }.bind(this))
                 .catch(err => {
-                    handleRequestError(err)
-                    reject()
+                    if( err ) {
+                        handleRequestError(err)
+                        reject()
+                    }
                 })
         })
     }

--- a/src/ItemsExplorer/SitemapPartialProvider.ts
+++ b/src/ItemsExplorer/SitemapPartialProvider.ts
@@ -13,7 +13,7 @@ const SNIPPET_TEMPLATE = (item: Item): SnippetString => {
     if (item.members.length) {
         let members = item.members.map(member => `Default item=${member.name} label="${member.label || ''}"`)
         snippet = `
-Text item="${item.name}" label="${item.label}" icon="${item.category || 'none'}" {
+Text item=${item.name} label="${item.label}" icon="${item.category || 'none'}" {
     ${members.join('\n    ')}
 }`
     }

--- a/src/ThingsExplorer/ThingsExplorer.ts
+++ b/src/ThingsExplorer/ThingsExplorer.ts
@@ -3,8 +3,7 @@ import {
     EventEmitter,
     TreeDataProvider,
     TreeItem,
-    TreeItemCollapsibleState,
-    workspace
+    TreeItemCollapsibleState
 } from 'vscode'
 
 import * as path from 'path'
@@ -72,8 +71,7 @@ export class ThingsExplorer implements TreeDataProvider<Thing|Channel> {
     }
 
     public getChildren(thing?: Thing): Thing[] | Thenable<Thing[]> | Channel[] | Thenable<Channel[]> {
-        let config = workspace.getConfiguration('openhab')
-        if (!thing && config.useRestApi) {
+        if (!thing) {
             if (!this.model) {
                 this.model = new ThingsModel(this.openhabHost)
             }

--- a/src/ThingsExplorer/ThingsModel.ts
+++ b/src/ThingsExplorer/ThingsModel.ts
@@ -59,7 +59,13 @@ export class ThingsModel {
 
     protected sort(nodes: Thing[]): Thing[] {
         return nodes.sort((n1, n2) => {
-            return n1.label.localeCompare(n2.label)
+            if (n1.label && n2.label) {
+                return n1.label.localeCompare(n2.label)
+            }
+
+            if (n1.UID && n2.UID) {
+                return n1.UID.localeCompare(n2.UID)
+            }
         });
     }
 }

--- a/src/ThingsExplorer/ThingsModel.ts
+++ b/src/ThingsExplorer/ThingsModel.ts
@@ -53,7 +53,6 @@ export class ThingsModel {
                 }.bind(this))
                 .catch(err => {
                     if( err ) {
-                        handleRequestError(err)
                         reject()
                     }
                 })

--- a/src/ThingsExplorer/ThingsModel.ts
+++ b/src/ThingsExplorer/ThingsModel.ts
@@ -35,7 +35,7 @@ export class ThingsModel {
      * Returns members of Group-type Item
      * @param thing openHAB root Item
      */
-    public getChildren(thing: Thing): Channel[] {
+    public getChildren(thing: Thing): Channel[] | any[] {
         return thing.channels ? thing.channels : []
     }
 
@@ -52,9 +52,7 @@ export class ThingsModel {
                     resolve(this.sort(transform(response)))
                 }.bind(this))
                 .catch(err => {
-                    if( err ) {
-                        reject()
-                    }
+                    handleRequestError(err).then(err => resolve([]))
                 })
         })
     }

--- a/src/ThingsExplorer/ThingsModel.ts
+++ b/src/ThingsExplorer/ThingsModel.ts
@@ -36,7 +36,7 @@ export class ThingsModel {
      * @param thing openHAB root Item
      */
     public getChildren(thing: Thing): Channel[] {
-        return thing.channels
+        return thing.channels ? thing.channels : []
     }
 
     private sendRequest(uri: string, transform): Thenable<Thing[]> {
@@ -52,8 +52,10 @@ export class ThingsModel {
                     resolve(this.sort(transform(response)))
                 }.bind(this))
                 .catch(err => {
-                    handleRequestError(err)
-                    reject()
+                    if( err ) {
+                        handleRequestError(err)
+                        reject()
+                    }
                 })
         })
     }

--- a/src/Utils.ts
+++ b/src/Utils.ts
@@ -121,7 +121,6 @@ export async function handleRequestError(err) {
     const result = await window.showErrorMessage('Error while connecting to openHAB REST API. ', setHost, disableRest)
     switch (result) {
         case setHost:
-            config.update('host', 'localhost')
             commands.executeCommand('workbench.action.openWorkspaceSettings')
             break
         case disableRest:

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -116,7 +116,6 @@ async function init(context: ExtensionContext, disposables: Disposable[]): Promi
     if (isOpenHABWorkspace()) {
         disposables.push(window.registerTreeDataProvider('openhabItems', itemsExplorer))
         disposables.push(window.registerTreeDataProvider('openhabThings', thingsExplorer))
-        disposables.push(languages.registerCompletionItemProvider('openhab', itemsCompletion))
 
         if (config.lspEnabled) {
             let languageClientProvider = new LanguageClientProvider()

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -112,22 +112,6 @@ async function init(context: ExtensionContext, disposables: Disposable[]): Promi
         return config.paperInBrowser ? openBrowser(route.replace(/%23/g, '#')) : openUI(options, title)
     }))
 
-    if (isOpenHABWorkspace()) {
-        if (config.useRestApi) {
-            disposables.push(window.registerTreeDataProvider('openhabItems', itemsExplorer))
-            disposables.push(window.registerTreeDataProvider('openhabThings', thingsExplorer))
-        }
-
-        if (config.lspEnabled) {
-            let languageClientProvider = new LanguageClientProvider()
-            disposables.push(languageClientProvider.connect())
-        }
-
-        if (config.useRestApi && config.restCompletions) {
-            disposables.push(languages.registerCompletionItemProvider('openhab', new ItemsCompletion(getHost())))
-        }
-    }
-
     disposables.push(commands.registerCommand('openhab.command.refreshEntry', (query) => {
         itemsExplorer.refresh()
         thingsExplorer.refresh()
@@ -160,6 +144,21 @@ async function init(context: ExtensionContext, disposables: Disposable[]): Promi
     disposables.push(commands.registerCommand('openhab.command.things.copyUID', (query) =>
     ncp.copy(query.UID || query.uid)))
 
+    if (isOpenHABWorkspace()) {
+        if (config.useRestApi) {
+            disposables.push(window.registerTreeDataProvider('openhabItems', itemsExplorer))
+            disposables.push(window.registerTreeDataProvider('openhabThings', thingsExplorer))
+        }
+
+        if (config.lspEnabled) {
+            let languageClientProvider = new LanguageClientProvider()
+            disposables.push(languageClientProvider.connect())
+        }
+
+        if (config.useRestApi && config.restCompletions) {
+            disposables.push(languages.registerCompletionItemProvider('openhab', new ItemsCompletion(getHost())))
+        }
+    }
 }
 export function activate(context: ExtensionContext) {
     const disposables: Disposable[] = [];

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -50,7 +50,6 @@ async function init(context: ExtensionContext, disposables: Disposable[]): Promi
     let config = workspace.getConfiguration('openhab')
     const itemsExplorer = new ItemsExplorer(getHost())
     const thingsExplorer = new ThingsExplorer(getHost())
-    const itemsCompletion = new ItemsCompletion(getHost())
 
     disposables.push(commands.registerCommand('openhab.basicUI', () => {
         let editor = window.activeTextEditor
@@ -114,15 +113,17 @@ async function init(context: ExtensionContext, disposables: Disposable[]): Promi
     }))
 
     if (isOpenHABWorkspace()) {
-        disposables.push(window.registerTreeDataProvider('openhabItems', itemsExplorer))
-        disposables.push(window.registerTreeDataProvider('openhabThings', thingsExplorer))
+        if (config.useRestApi) {
+            disposables.push(window.registerTreeDataProvider('openhabItems', itemsExplorer))
+            disposables.push(window.registerTreeDataProvider('openhabThings', thingsExplorer))
+        }
 
         if (config.lspEnabled) {
             let languageClientProvider = new LanguageClientProvider()
             disposables.push(languageClientProvider.connect())
         }
 
-        if (config.restCompletions) {
+        if (config.useRestApi && config.restCompletions) {
             disposables.push(languages.registerCompletionItemProvider('openhab', new ItemsCompletion(getHost())))
         }
     }

